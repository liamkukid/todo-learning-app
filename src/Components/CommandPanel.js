import { memo } from 'react';
import PropTypes from 'prop-types';
import { FILTER_ACTIVE, FILTER_COMPLETED, SHOW_ALL } from '../Filters';

import style from './commandPanel.module.scss';

function CommandPanel({ todos, onFilterChanged, clearCompleted }) {
  return (
    <div className={style.panel}>
      <span className={style.span_items_left}>
        {todos !== null &&
          todos.length > 0 &&
          todos.filter(todo => !todo.done).length}{' '}
        items left
      </span>
      <Button title="All" onClick={() => onFilterChanged(SHOW_ALL)} />
      <Button title="Active" onClick={() => onFilterChanged(FILTER_ACTIVE)} />
      <Button
        title="Completed"
        onClick={() => onFilterChanged(FILTER_COMPLETED)}
      />
      <Button title="Clear Completed" onClick={clearCompleted} />
    </div>
  );
}

function Button({ title, onClick }) {
  return (
    <button
      className={style.button}
      type="button"
      onClick={onClick}
      value={title}
    >
      {title}
    </button>
  );
}

export default memo(CommandPanel);

CommandPanel.propTypes = {
  todos: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      done: PropTypes.bool,
      id: PropTypes.string,
    })
  ),
  onFilterChanged: PropTypes.func.isRequired,
  clearCompleted: PropTypes.func.isRequired,
};

CommandPanel.defaultProps = {
  todos: [],
};

Button.propTypes = {
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};
