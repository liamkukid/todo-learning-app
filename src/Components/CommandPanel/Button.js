import PropTypes from 'prop-types';
import style from './button.module.scss';

export default function Button({ title, onClick }) {
  return (
    <button
      className={style.button}
      type="button"
      onClick={onClick}
      value={title}
    >
      {title}
    </button>
  );
}
Button.propTypes = {
  title: PropTypes.string.isRequired,
  onClick: PropTypes.func.isRequired,
};
