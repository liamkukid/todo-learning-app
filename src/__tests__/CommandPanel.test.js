import renderer from 'react-test-renderer';
import { render, screen, fireEvent } from '@testing-library/react';
import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import CommandPanel from '../Components/CommandPanel/CommandPanel';
import todosReducer from '../features/todosSlice';
import filterReducer from '../features/filterSlice';

it('renders correctly with no todos', () => {
  const store = configureStore({
    reducer: {
      todos: todosReducer,
    },
    preloadedState: {
      todos: [],
    },
  });
  const tree = renderer
    .create(
      <Provider store={store}>
        <CommandPanel />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('renders correctly with only one todo done', () => {
  const store = configureStore({
    reducer: {
      todos: todosReducer,
    },
    preloadedState: {
      todos: [
        {
          title: 'todo 1',
          done: true,
          id: 'todo_1_id',
        },
        {
          title: 'todo 2',
          done: false,
          id: 'todo_2_id',
        },
      ],
    },
  });
  const tree = renderer
    .create(
      <Provider store={store}>
        <CommandPanel />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

describe('CommandPanel', () => {
  let store;

  beforeEach(() => {
    store = configureStore({
      reducer: {
        todos: todosReducer,
        filter: filterReducer,
      },
      preloadedState: {
        todos: [
          {
            title: 'todo 1',
            done: true,
            id: 'todo_1_id',
          },
          {
            title: 'todo 2',
            done: false,
            id: 'todo_2_id',
          },
        ],
        filter: 'no_filter',
      },
    });
  });

  afterEach(() => {
    store = null;
  });

  it('set filter to showAll', () => {
    render(
      <Provider store={store}>
        <CommandPanel />
      </Provider>
    );
    const linkElementButton = screen.getByText('All');

    fireEvent.click(linkElementButton);

    const state = store.getState();
    expect(state.filter).toEqual('SHOW_ALL');
  });

  it('set filter to filterActive', () => {
    render(
      <Provider store={store}>
        <CommandPanel />
      </Provider>
    );
    const linkElementButton = screen.getByText('Active');

    fireEvent.click(linkElementButton);

    const state = store.getState();
    expect(state.filter).toEqual('FILTER_ACTIVE');
  });

  it('set filter to filterCompleted', () => {
    render(
      <Provider store={store}>
        <CommandPanel />
      </Provider>
    );
    const linkElementButton = screen.getByText('Completed');

    fireEvent.click(linkElementButton);

    const state = store.getState();
    expect(state.filter).toEqual('FILTER_COMPLETED');
  });

  it('remove completed todos', () => {
    render(
      <Provider store={store}>
        <CommandPanel />
      </Provider>
    );
    const linkElementButton = screen.getByText('Remove Completed');

    fireEvent.click(linkElementButton);

    const state = store.getState();
    expect(state.todos.length).toEqual(1);
    expect(state.todos[0].id).toEqual('todo_2_id');
  });
});
