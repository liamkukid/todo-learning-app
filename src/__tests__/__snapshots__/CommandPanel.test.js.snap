// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no todos 1`] = `
<div
  className="panel"
>
  <span
    className="span_items_left"
  >
    no
     items left
  </span>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="All"
  >
    All
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="Active"
  >
    Active
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="Completed"
  >
    Completed
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="Remove Completed"
  >
    Remove Completed
  </button>
</div>
`;

exports[`renders correctly with only one todo done 1`] = `
<div
  className="panel"
>
  <span
    className="span_items_left"
  >
    1
     items left
  </span>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="All"
  >
    All
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="Active"
  >
    Active
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="Completed"
  >
    Completed
  </button>
  <button
    className="button"
    onClick={[Function]}
    type="button"
    value="Remove Completed"
  >
    Remove Completed
  </button>
</div>
`;
